Improvements to instance search performance

plt-amy opened the pull request:
A bunch of minor improvements to instance search performance, with — ideally! — no impact to correctness. Each commit has a description of the individual change.

The big one is 4f80b0ce86ef029e59a148b658abab59a1f69ddf, which essentially undoes #7298, since 94b1d277e540988f6674218410fe1f1cb459a7df provides an alternate fix to #7295, by making there be no overlap instead (at the use-site, the instance constraint is perfectly unambiguous, but the discrimination tree thought that e.g. `Nat-preorder` had type `Has-Preorder _ _` *in addition to* its actual type).

As described in #7298, the logic to postpone an instance early is... well, fiddly: to preserve backwards compatibility, there's a lot of asterisks as for when we can turn "the discrimination tree came up with multiple results" into "we should wait on a given meta". The big ones are:

* To preserve the "inert improvement" behaviour, it needs to be possible to turn off this postponement at all. In the test, we have a constraint `Monad _M`, with the only constraint that could solve `_M` being a blocked `_M Nat := State Nat Nat`, and the only way to make progress is to test *all* the possible `Monad` instances until we find the right one. Having to try all the candidates sucks extremely, but we have to preserve this behaviour.

* To work around #7186, we can't postpone when one of the instances belongs to the same mutual block we're checking under, since if we wait for too long, the instance will be solved with a definition instead of a record literal, and this makes the termination checker very angry.

There are some more conditions, like e.g. we shouldn't block a macro that just wants to get all the instances, but those are the big ones. As described in 4f80b0ce86ef029e59a148b658abab59a1f69ddf's message, in a particular 1Lab module ([here](https://github.com/the1lab/1lab/blob/aliao/elementary-topoi/src/Cat/Instances/Presheaf/Subobjects.lagda.md#L93-L110)) we use a bunch of instances (of e.g. `Extensional`) in a where clause against a meta that's only solved by the RHS of the parent clause. On master, this has the pathological behaviour of repeatedly retrying every possible `Extensional` instance against this meta, resulting in something like 280 candidate checks that would be immediately discarded by the discrimination tree. With these patches, we only check these metas once, which nets a 2s improvement to TC time.

[Here](https://gist.github.com/plt-amy/6cbf648bf49589266094343f9f5ed566) are logs of `-vtc.instance:30` both with the "fiddly postponement" (`subobj-out`) and without (`subobj-out-lazy`); note how `subobj-out-lazy` has an extra 9000 lines of attempts for `Extensional` instances, and the difference in type-checking time at the bottom of both files (though this is *greatly* exaggerated by the debug printing; in the actual 1Lab the file took 6s to check before and 3 after).

P.S.: Andreas: closes #7486.

plt-amy committed:
Treat opaque defs rigidly when finding instances

This prevents Agda from trying *every* instance of a class, when looking
for an instance for an opaque thing that is not being unfolded in the
current context.

plt-amy committed:
Record Pi-type visibility in instance table

In the 1Lab we have instances

  _ : C ({x : A} → B x)
  _ : C ((x : A) → B x)
  _ : C (⦃ x : A ⦄ → B x)

and prior to this patch, all of them would be tried every time, even
though the visibility of a pi type is easy to record properly.

plt-amy committed:
Overwrite instance table entry when adding an alias

Less violent fix for #7295. The issue there is that we'd add each
instance twice, once with a useful type, and once with a bunch of metas.
So the discrimination tree would look like

  case @0 of
    Has-Preorder² → case @0 of
      Nat → done {Nat-preorder}
      ...
    * → done {Nat-preorder, Bool-preorder}

which, to initialInstanceCandidates, looks like we're dealing with
actual overlap.

plt-amy committed:
Block instances on discrimination tree overlap

Fixes a performance regression introduced by #7298 to the 1Lab. The
issue boils down to a constraint like

    _0 : Extensional ∣ ₀ _P_669 y ∣ _ℓr_724

where the bars and ₀ are both projections. In the module which prompted
this fix, there are *29* total 'Extensional' instances, and, of course,
they would all be tried for this constraint. Moreover, since none of the
candidates can be rejected until _P_669 is solved, this constraint would
be retried over and over, resulting in

    % grep 'checkCandidateForMeta _725@' [debug output] | wc -l
    260

quite a few useless retries! With this patch, we wait until _P makes
progress, so only *one* candidate is checked for this meta.

In this module, the time spent in checkCandidateForMeta is cut by 50%.

plt-amy commented:
The stdlib failure (which is the same as #7450) is because `dropSameCandidate` uses a syntactic check to make sure that the candidate doesn't mention any metas. On `master`, adding the instance constraint *with the candidates* (which is what we're trying to avoid, because all the candidates in the constraint get re-checked) causes all the candidates to get reduced, so we get:

```
checkCandidateForMeta
  t    = Foreign.Haskell.Coerce.Coercible {zero} {zero}
         (Maybe {zero} (FilePath n)) (Maybe {zero} (FilePath n))
  t'   = {a b : Level} →
         Foreign.Haskell.Coerce.Coercible₁ {a} {b} a b (Maybe {a})
         (Maybe {b})
  term = λ {_} {_} {_} {_} ⦃ _ ⦄ → Foreign.Haskell.Coerce.TrustMe
[...]
instance search: found solution for _716:
  Foreign.Haskell.Coerce.TrustMe
```

which does not mention any new metas when applied, but after this patch we get:

```
{ checkCandidateForMeta _767@4831777243146177906
checkCandidateForMeta
  t    = Foreign.Haskell.Coerce.Coercible {zero} {zero}
         (Maybe {zero} (FilePath n)) (Maybe {zero} (FilePath n))
  t'   = {a b : Level} →
         Foreign.Haskell.Coerce.Coercible₁ {a} {b} a b (Maybe {a})
         (Maybe {b})
  term = Foreign.Haskell.Coerce.coerce-maybe
[...]
instance search: found solution for _767:
  Foreign.Haskell.Coerce.coerce-maybe {zero} {zero} {FilePath n}
  {FilePath n} ⦃ _782 ⦄
[...]
dropSameCandidates: Solution of instance meta has fresh metas so we don't filter equal candidates yet
```

plt-amy committed:
fixup: reduce head instance before dropping dupes

Fixes an issue where the standard library subtly relies on the fact that
instance candidates get reduced when an instance constraint gets
postponed to implement overlapping behaviour. More specifically, they
have

    coerce-maybe : ⦃ Coercible A B ⦄ → Coercible (Maybe A) (Maybe B)
    coerce-maybe _ = TrustMe

    coerce-refl : Coerce A A
    coerce-refl = TrustMe

and for the code to type-check we *have* to reduce coerce-maybe to a
lambda, otherwise we don't have TrustMe ≠ TrustMe to "resolve" the
overlap.

plt-amy commented:
For completeness, here's a CI run of the 1Lab with these patches: https://github.com/the1lab/1lab/actions/runs/13597394855/job/38017234881

Edit: 2 entire minutes faster than the last [full rebuild](https://github.com/the1lab/1lab/actions/runs/12948014985/job/36115775079) >:)

andreasabel commented:
I am not proficient with the instance search so I can only give comments from the very outside:

> which essentially undoes #7298

This was released in 2.7.0, so I'd just wanted to ask the routine question of whether we expect any regressions from the current PR.
There is one regression reported for 2.7.0, is this related, and would it be fixed by this PR?
- #7486

plt-amy commented:
I don't *think* so? It's possible that some code out there relies on the "we're going to try literally every instance" behavior, e.g. in a `with` clause, but that would be easy to fix by enabling `stFinalChecks` while checking the `with` scrutinee (and probably giving the field a more derogatory name like `stFragileInstancesContext` or something) — and this PR would still be a performance win then, as long as that user's code doesn't mostly exist *in* `with` abstractions.

As for #7486, I stand by our decision to brand that issue a bugfix, so I guess I can add the test cases from there (both the now-failing program and the now-succeeding program) in this PR.

plt-amy committed:
Test cases for #7486

Treats the new behaviour (instances for a reducible function are
wildcardsa) as correct, and adds the previously-accepted, now-failing
programs that would silently ignore all the possible instances as
failing test cases.

Adds a successful test case for adding a pair of instances at once by
branching on a boolean in the return type.

andreasabel commented:
> As for #7486, I stand by our decision to brand that issue a bugfix, so I guess I can add the test cases from there

Ok, I added this issue to this PR.

plt-amy commented:
```
Acquired cache lock on /home/runner/.stack/pantry/hackage/hackage-security-lock
Released cache lock on /home/runner/.stack/pantry/hackage/hackage-security-lock
Verification loop. Errors in order:
  Invalid hash for <repo>/snapshot.json
  Invalid hash for <repo>/snapshot.json
  Invalid hash for <repo>/snapshot.json
  Invalid hash for <repo>/snapshot.json
  Invalid hash for <repo>/snapshot.json
```
I don't think the size solver failure is this PR's fault :sweat_smile:
> Ok, I added this issue to this PR.

Thanks!

jespercockx commented:
Thanks for this PR and the in-depth analysis of Agda's admittedly mysterious postponing behavior.

> On master, adding the instance constraint with the candidates (which is what we're trying to avoid, because all the candidates in the constraint get re-checked) causes all the candidates to get reduced

I believe there is no good reason for why we actually reduce the candidates (rather than just the type), and in fact it would be better not to reduce them. See https://github.com/agda/agda/issues/7449

plt-amy commented:
I don't think there's a sensible way to keep `dropSameCandidates` without reducing (really, we should be *normalising*) the candidates — it's not safe to compare two terms in the "ambient" metacontext that mention "future" metavariables (i.e. metavariables that were invented during `checkCandidateForMeta`, and so will only exist to the rest of the program when we actually choose a candidate), and reducing (or not) influences whether these appear in the resulting term.

And the standard library relies on this behaviour:

```haskell
-- This rule overlaps with the purely structural ones: when attempting
-- to prove `Coercible (List A) (List A)`, should Agda use the proof
-- obtained by `coerce-refl` or the one obtained by `coerce-list coerce-refl`?
-- Because we are using a datatype with a single constructor these
-- distinctions do not matter: both proofs are definitionally equal.
```

plt-amy commented:
Personally, I think `dropSameCandidates` is a hell function and we should get rid of it now that there are overlap pragmas — but the standard library also relies on us arbitrarily choosing the first candidate for an irrelevant meta to solve *non-instance* metas in code that uses `NonZero` over the naturals.

JacquesCarette commented:
> but the standard library also relies on us arbitrarily choosing the first candidate for an irrelevant meta to solve non-instance metas in code that uses `NonZero` over the naturals.

If you can make a solid case that this is bad practice, we can change the standard library. I wouldn't want the standard library to be the reason to keep a bad design around. [Note that I'm not the biggest user of `NonZero` so I'm not really the one who may need the most convincing, though I'm willing to slog through fixing things if that's what it takes.]

plt-amy commented:
Well, I think both that *and* whether we should reduce instance candidates is besides the point of whether we should merge this PR — I had to add the reduce to preserve the behaviour of treating `Coercible` instances as interchangeable, but IMO the standard library should just mark the base `coerce-refl` instance `INCOHERENT` — but I did end up writing an entire thing about how the "irrelevant instance meta → choose first instance" behaviour is fragile, so here goes:

---

@JacquesCarette It's a fragile design that relies on the implementation details of instance search to work around a lack of injectivity in certain `Nat` functions (the examples I remember are `_%_` and `_/_`) — which heavily restricts the ways we could change how the implementation of instance search orders the initial list of candidates, or how they're considered more generally. For a concrete example, we can look at how #7373 failed, which, though the logs are gone, I remember! It's `Data.Nat.DivMod`, and indeed replaying that patch we can see that this function breaks:

```agda
m∣n⇒o%n%m≡o%m : ∀ m n o .⦃ _ : NonZero m ⦄ .⦃ _ : NonZero n ⦄ → m ∣ n →
                o % n % m ≡ o % m
m∣n⇒o%n%m≡o%m m n@.(p * m) o (divides-refl p) = begin-equality
  o % n % m                ≡⟨⟩
  o % pm % m               ≡⟨ %-congˡ (m%n≡m∸m/n*n o pm) ⟩ -- breaks here
  (o ∸ o / pm * pm) % m    ≡⟨ cong (λ # → (o ∸ #) % m) (*-assoc (o / pm) p m) ⟨
  (o ∸ o / pm * p * m) % m ≡⟨ m*n≤o⇒[o∸m*n]%n≡o%n (o / pm * p) lem ⟩
  o % m                    ∎
  -- there's more stuff here ...
```

That's because `%` is not injective enough to solve for the `{o = m}` argument to `%-congˡ`, and indeed adding that annotation fixes this function the #7373 patch. On `master`, this works because Agda puts candidates from the context before candidates from the signature, so when we want to solve for `NonZero _o`, we arbitrarily choose `NonZero m` rather than the declared `NonZero (suc _)` instance, and this accidentally gets us the right answer! But if one day I woke up feeling zany and applied this patch on top of master

```diff
      reportSLn "tc.instance" 30 "dropSameCandidates: Meta is irrelevant so any candidate will do."
-     return [cvd]
+     return [last cands]
```

since *any candidate will do*, this should work; but now we fail with unsolved constraints instead:

```
  suc _n_308 = _o_300 : ℕ (blocked on _o_300)
  (o ∸ o / pm * pm) % _o_300 = (o ∸ o / pm * pm) % m : ℕ
    (blocked on _o_300)
  o % pm % _o_300 = o % pm % m : ℕ (blocked on _o_300)
  ```

The solution *here* is to make the `o` argument to `%-congˡ` explicit, since it can't be inferred in general — we can't solve `_x` from a constraint `m % _x = m % y`. The function `/-congˡ` is suspicious for the same reason, and now fails on lines 308, 348, 357, 358, 377, and 409; There's a yellow `/+-distrib-/-∣ʳ` on 360.

JacquesCarette commented:
Thanks for the very thorough description. I will open an issue on `stdlib` that refers back to this. In particular, it will explicitly mention the need to change `%-congˡ`and other like functions to take more arguments explicitly.

plt-amy commented:
@jespercockx Am I good to merge this, and then we can address the issue of reducing instance candidates at a later date (perhaps when the standard library makes explicit its intent to have an overlappable base instance)?

jespercockx commented:
Yes, I have no objections to merging this as it is.
